app/models/event.rb

```diff
@@ -118,6 +121,48 @@ class Event < ApplicationRecord
       .where("flipper_gates.feature_key = ? AND flipper_gates.key = ?", flag, "actors")
   }
 
+  def ancestor_ids
+    [id] + Event.connection.execute(<<-SQL).map { |row| row["id"] }
+      WITH RECURSIVE parent_events AS (
+        SELECT id, parent_id
+        FROM events
+        WHERE id = #{id}
+        UNION ALL
+        SELECT e.id, e.parent_id
+        FROM events e
+        INNER JOIN parent_events pe ON e.id = pe.parent_id
+      )
+      SELECT id FROM parent_events WHERE id != #{id};
+    SQL
+  end
+
+  def descendant_ids
+    Event.connection.execute(<<-SQL).map { |row| row["id"] }
+      WITH RECURSIVE child_events AS (
+        SELECT id, parent_id
+        FROM events
+        WHERE parent_id = #{id}
+        UNION ALL
+        SELECT e.id, e.parent_id
+        FROM events e
+        INNER JOIN child_events ce ON e.parent_id = ce.id
+      )
+      SELECT id FROM child_events;
+    SQL
+  end
+
+  def ancestors
+    Event.where(id: ancestor_ids)
+  end
```

Comment:
if there's any way we could make this a scope that would be much appreciated: https://x.com/nateberkopec/status/1907240790725890074.

Reply:
```ruby
scope :ancestors_of, -> (event){ 
      find_by_sql(
        ["WITH RECURSIVE parent_events AS (
          SELECT id, parent_id
          FROM events
          WHERE id = ?
          UNION ALL
          SELECT e.id, e.parent_id
          FROM events e
          INNER JOIN parent_events pe ON e.id = pe.parent_id
        )
        SELECT DISTINCT events.* from events WHERE id IN (SELECT id FROM parent_events)",
        event.id
      ]
    )
  }
```